### ISSUE TITLE
INTEGRAL() behaviour in GROUP BY

### ISSUE THREAD

**Issue by mzanders** [influxdata/influxdb#13164]
**System info**
InfluxDB 1.7.4 and 1.7.5 - Linux x64

**Steps to reproduce**
```
create database db
use db
insert test value=1 500000000
insert test value=1 1500000000
insert test value=1 2500000000
insert test value=1 3500000000
insert test value=1 4500000000
insert test value=1 5500000000
insert test value=1 6500000000
insert test value=1 7500000000
SELECT integral("value") FROM "test" WHERE time>2000000000 GROUP BY time(2s)
drop database db
```
**Expected behaviour**
```
name: test
time       integral
----       --------
2000000000 2
4000000000 2
6000000000 1.5
```
**Actual behaviour**
```
name: test
time       integral
----       --------
2000000000 1.5
4000000000 2
6000000000 1.5
```
**Additional info**
This type of query is a typical use case for integrating power into energy or counting 'on state' like in this simplified example. The problem is not so much apparent when the GROUP BY interval is small compared to the query input data (ie only the first aggregated data point is missing some 'area').
However, when used in a CQ, the dataset is reduced. In a simple CQ, only the last interval (2s) is used and stored which is obviously incorrect as the data is not complete after now() yet. Increasing the CQ dataset with RESAMPLE FOR 6s yield results as observed above. 
The first point in the first window is missing it's predecessor in the dataset and hence no integration can occur for that point, so the integral of the first fraction is missing and the total integral for the window is 'incomplete'. Upon the next CQ run, the windows move on 2s and the result that was correct in the previous run (2s) gets overwritten by an incorrect value again.
The end result of this is that the integral is missing amplitude, depending on the misalignment between source data and aggregation period.
You could think of it as a fancy way to generate saw-tooth waveforms by using constant data and slightly offset sample rates.

**Comment by mzanders** [comment]
#10289

**Comment by mzanders** [comment]
#10289 Can partly be declared by this behaviour I think. (although not the 0-points)
#9645 Is also in the same area.

**Comment by mzanders** [comment]
I have implemented an improvement in the integral functions:
https://github.com/influxdata/influxdb/compare/v1.7.5...Tiger-Power:integral_groupby_impr

The idea is that:
-  when there's no GROUP BY, the reducer works as before
-  when there is a GROUP BY, the reducer only outputs data for windows which have direct predecessor and successor points in the adjoining windows

So for the test case above, it results in:
```
name: test
time       integral
----       --------
4000000000 2
```

As you see, this change has a rather dramatic impact on the output depending on the input data:
-  it requires a valid point in every window, in other words for (almost) constant sample rate data: the group by interval needs to be large enough to always host 1 sample.
-  the output data range doesn't match with the query time range, which might seem strange. In fact, the first and last window will be missing by default. Mathematically these were incorrect anyway (and there's no way to go about that as we're limited to the dataset of the query). 
-  You could discuss about the last window (which gets overwritten in a CQ anyway). Here, I prefer to alert the user that he is calculating erroneous data and not output anything. This prevents the user from happily jotting down data in a CQ output which is wrong.
-  when there's no data for any number of windows in the middle of the data set, the windows before and after that gap will also get cropped. This is solving #9645 at the expense of removing the output of those neighbouring windows. But here again, there's no way to correctly calculate the integral for that window anyway.

For sure, when I now run a sum() on the aggregated integral after 24hrs of data I'm getting almost the same result as the integral on the original data. *Almost* as the aggregation is always lagging by a full GROUP BY window and the normal integral query (no group by) is (still) missing the area up to the first point.
I'm rather convinced that this is the best mathematical solution for the problem but like to have some feedback from people who have been using Influx for more than 1 week.
@oori  @roberthartung @rbetts @jsternberg?

**Comment by mzanders** [comment]
Thinking about the time reported by INTEGRAL(), it's not very logical to put this either at the start or the end of the window, is it? It should always be in the middle of the window, regardless of direction.
Ie if you would overlap columns with integrated data on the source data, you'd want these to align time-wise.

**Comment by mzanders** [comment]
Can someone elaborate these lines? They're not present in the float version, but are in the int and unsigned int. No similar check present for descending order.
```
		// If we see the minimum allowable time, set the time to zero so we don't
		// break the default returned time for aggregate queries without times.
		if r.window.start == influxql.MinTime {
			r.window.start = 0
		}
```

**Comment by mzanders** [comment]
https://github.com/Tiger-Power/influxdb/commit/227d101d12c3f7184c2a5baa9945c660446ad236
This commit fixes #10289
`> select integral("value",1h) from test group by time(5m)
name: test
time                integral
----                --------
1554365100000000000 -212.45654449074073
1554365400000000000 -213.9228785185185
1554365700000000000 -220.67323337962964
1554366000000000000 -218.3902332175926
1554366300000000000 -215.05824969907408
1554366600000000000 -213.00527333333332
1554366900000000000 -209.47371833333332
1554367200000000000 -217.14359208333332
1554367500000000000 -210.08914837962965
1554367800000000000 -218.62753500000002
1554368100000000000 -220.79359555555558
1554368400000000000 -204.9215363425926
1554368700000000000 -212.71889166666668
> select integral("value",1h) from test group by time(5m) order by time DESC
name: test
time                integral
----                --------
1554368700000000000 -212.71889166666668
1554368400000000000 -204.9215363425926
1554368100000000000 -220.79359555555556
1554367800000000000 -218.62753500000002
1554367500000000000 -210.08914837962965
1554367200000000000 -217.14359208333332
1554366900000000000 -209.47371833333335
1554366600000000000 -213.0052733333333
1554366300000000000 -215.05824969907408
1554366000000000000 -218.3902332175926
1554365700000000000 -220.67323337962964
1554365400000000000 -213.9228785185185
1554365100000000000 -212.4565444907407
`
